scRNAseq_and_smFISH

maxime1310 opened the pull request.

jeff-regier commented on scvi/dataset/data_loaders.py:
```diff
@@ -190,3 +190,20 @@ def classifier_data_loaders(self):
+        fish = TrainTestDataLoaders(gene_dataset_fish, train_size, test_size, seed, **data_loaders_kwargs)
+        self.dict = {}
+        self.dict.update({'train_seq': seq['train'], 'test_seq': seq['test'],
+                         'train_fish': fish['train'], 'test_fish': fish['test']})
```
Looks like we should be able to just use the normal `TrainTtsetDataLoaders` here.

jeff-regier commented on scvi/dataset/dataset.py:
```diff
@@ -53,10 +57,18 @@ def download_and_preprocess(self):
-            torch.FloatTensor(self.local_vars[indexes]), \
-            torch.LongTensor(self.batch_indices[indexes]), \
-            torch.LongTensor(self.labels[indexes])
+        if not self.got_spatial:
```
Let's just have `if x_coord is not None:` here, and get rid of the `got_spatial` variable.

jeff-regier commented on scvi/dataset/dataset.py:
```diff
@@ -32,6 +32,10 @@ def __init__(self, X, local_means, local_vars, batch_indices, labels,
+        if x_coord is not None:
+            self.got_spatial = True
+        else:
+            self.got_spatial = False
```
Let's get rid of the `got_spatial` variable, and just use `x_coord not is None` if we need it. Better to have fewer variables.

jeff-regier commented on scvi/dataset/dataset.py:
```diff
@@ -53,10 +57,18 @@ def download_and_preprocess(self):
+            return X, torch.FloatTensor(self.local_means[indexes]), \
+                   torch.FloatTensor(self.local_vars[indexes]), \
+                   torch.LongTensor(self.batch_indices[indexes]), \
+                   torch.LongTensor(self.labels[indexes])
```
Better to return dummy values for x_coord and y_coord, like we do for batch_indices when there aren't batches, rather than branching

jeff-regier replied:
@Edouard360  Could we move `collate_fn` to the `Inference` classes? (and override it in `VariationalInferenceFish` with this one?)

For that matter, we move all the `DataLoaders` logic (e.g. `TrainTestDataLoader`) to the `Inference` class? It seems like almost every Inference class e.g. (`VariationalInference`) needs a different data loader any way -- it'd be nice not to have two separate hierarchies of classes.

Edouard360 replied:
`TrainTestDataLoader` is also used in `ClassifierInference`, which serves to: 1) the "alternate" training scheme `SemiSupervisedModel` in which at the end of each epoch, we train only the classifier weights on top of the latent space 2) if we wanted to do a NN classifier on top of a VAE or on raw data.

I see what you mean. It is possible to move the functionalities of `DataLoaders`  into `Inference`, only keeping the `DataLoadersWrapper` class.

`collate_fn` is a  method of pytorch's `Dataset` class used by pytorch's dataloaders however... I am not sure to understand how to make it part of Inference.

jeff-regier commented on scvi/dataset/smfish.py:
```diff
@@ -18,7 +18,7 @@ def __init__(self, save_path='data/'):
 
     def preprocess(self):
-        print("Preprocessing smFISH dataset")
+        print("Preprocessing smFISH_utils dataset")
```
What utils?

jeff-regier commented on scvi/inference/variational_inference.py:
```diff
@@ -306,3 +310,103 @@ def loss(self, tensors_all, tensors_labelled):
+    def loss(self, tensors_seq, tensors_fish):
+        sample_batch, local_l_mean, local_l_var, batch_index, labels = tensors_seq
+        reconst_loss, kl_divergence = self.model(sample_batch, local_l_mean, local_l_var, batch_index, mode="scRNA",
+                                                 ponderation=self.ponderation)
```
ponderation -> weighting ?

jeff-regier commented on scvi/inference/variational_inference.py:
```diff
@@ -306,3 +310,103 @@ def loss(self, tensors_all, tensors_labelled):
+        loss /= (sample_batch.size(0) + sample_batch_fish.size(0))
+        return loss + loss_fish
+
+    def train(self, n_epochs=20, lr=1e-3, params=None):
```
Let's find a way to use the `train` method from `VariationalInference` rather than overriding it.

jeff-regier commented on scvi/metrics/clustering.py:
```diff
@@ -74,3 +75,52 @@ def entropy(hist_data):
     return score / 50
+
+
+def get_data(vae, data_loader, mode):
```
Every variable is data. Let's give this function a name that will help someone who is reading the code for the first time to understand what it does.

jeff-regier commented on scvi/metrics/log_likelihood.py:
```diff
@@ -4,13 +4,18 @@
 
 
-def compute_log_likelihood(vae, data_loader):
+def compute_log_likelihood(vae, data_loader, name):
```
Better not to take `name` as an argument. Instead, let's make `compute_log_likelihood`  a method in the `VariationalInference` class. Then override `compute_log_likelihood` in the `VariationalInferenceFish` class.

jeff-regier commented on smFISH_utils/metrics/visualisation.py:
```diff
@@ -0,0 +1,228 @@
+
+
+def show_cell_types(latent_seq, labels_seq, latent_fish, labels_fish, title='latent.svg', x_lim=[-85, 85],
+                    y_lim=[-85, 85]):
```
Would it work to make all these smfish-specific function methods in the `VariationalInferenceFish` class?

jeff-regier replied:
That would keep all the smFISH-specific code together in one place.

maxime1310 commented:
@jeff-regier we just went through all the new code with Ed, he has a lot of ideas on how to factorize it (I don't know how I would survive without him haha), we'll probably be able to get to a much cleaner version.

jeff-regier commented on smFISH/metrics/visualisation.py:
```diff
@@ -0,0 +1,228 @@
+import matplotlib.pyplot as plt
```
I'm not sure I understand -- maybe just make whatever changes you think are good and let me look over the code before committing it?

I'm interested in getting rid of everything in the `metrics` directory eventually, and moving it all to `Inference` or `VariationalInference`. Currently, it seems like every metric is defined two places --- in an `Inference` class (or subclass of it) and then again in the `metrics` directory. Seems unnecessarily complicated.

But I could easily be missing something about the importance of the `metrics` directory -- glad to see what you can do to simplify it.

Edouard360 replied:
Shouldn't all this metrics be separated in different files ? If so would you prefer to move them all in Inference directory ? Or if not, this would mean the `variational_inference.py` file would get pretty big, with all the methods following one another. 

It is true that very few functions would be used elsewhere - there are the clustering scores that might serve for other algorithms.

jeff-regier replied:
I'm not sure having long files is a problem.

jeff-regier commented on scvi/metrics/classification.py:
```diff
@@ -84,29 +86,35 @@ def compute_accuracy_svc(data_train, labels_train, data_test, labels_test, unit_
+                compute_accuracy_tuple(labels_test, y_pred_test))
+    elif return_labels:
+        return (compute_accuracy_tuple(labels_train, y_pred_train),
+                compute_accuracy_tuple(labels_test, y_pred_test), y_pred_test)
```
Would you just always return the labels, rather than having this if statement? If you call the function and you don't need them, you can just use and underscore as a variable to assign to; e.g.,
```
train_acc, test_acc, _ = infer.compute_accuracy_classifier(...)
```

maxime1310 replied:
I'll do that and modify accordingly the part of the code where we call that function.

jeff-regier commented on scvi/metrics/log_likelihood.py:
```diff
@@ -63,4 +49,5 @@ def log_nb_positive(x, mu, theta, eps=1e-8):
     res = theta * torch.log(theta + eps) - theta * torch.log(theta + mu + eps) + x * torch.log(
         mu + eps) - x * torch.log(theta + mu + eps) + torch.lgamma(x + theta) - torch.lgamma(
         theta) - torch.lgamma(x + 1)
+    # Gives the ability to do a warmup on the unobserved smFISH genes when merging the two technologies
```
How so? Is this comment outdated?

maxime1310 replied:
Yes it's outdated! I'm currently moving the previous logic of weighting some genes more than other within the vae_fish model

jeff-regier commented on scvi/models/vae_fish.py:
```diff
@@ -0,0 +1,202 @@
+        self.reconstruction_loss_fish = reconstruction_loss_fish
+        self.model_library = model_library
+        # First layer of the encoder isn't shared
+        self.z_encoder = Encoder(n_input, n_hidden, n_hidden=n_hidden, n_layers=1,
```
Is there a `z_encoder` already in the parent class? If so, you wouldn't want to instantiate a new one here, I suspect.

maxime1310 replied:
There is one but it doesn't take the same parameters as input, so if we use the one from the parent class we'd have to specify different parameters in the init (and then we'd lose the interest of inheriting from VAE as we'd have to respecify most features for the initialization)

jeff-regier replied:
You mean in the call to `super(VAEF, self).__init__(...)` you'd have to specify some different parameters? That's OK --it's an improvement over how it is now. Currently we're getting instantiating `z_encoder` twice, but the parent instance is masked so it's never seen. That's inefficient and confusing for someone new to the code. e.g., if some moves the call to `super(VAEF, self).__init__(...)`  to the end of the `__init__` function, it would totally change the behavior.

jeff-regier commented on scvi/dataset/dataset.py:
```diff
@@ -255,6 +265,24 @@ def _filter_genes(gene_dataset, gene_names_ref, on='gene_names'):
         return gene_dataset.X[:, subset_genes], subset_genes
 
+    @staticmethod
+    def reorganize(x, genes, ordered_genes):
```
It looks like this is only used in `CortexDataset`. Please move it there to keep it near where it's used. 

Also, please give a name that's a bit more descriptive than `reorganize`. Maybe `reorder_genes` or something?

maxime1310 replied:
I'll move that and rename it. I really need to start naming my methods in a more informative way, I always come up with lousy names 😞

jeff-regier commented on scvi/inference/experimental_inference.py:
```diff
@@ -73,6 +73,22 @@ def adversarial_loss(self, tensors, *next_tensors):
     return type(self).loss(self, tensors, *next_tensors) - cls_loss
 
 
+def adversarial_loss_fish(self, tensors_seq, tensors_fish):
```
Is it possible to put these smFISH-specific methods all the `VariationalInferenceFish` class? It'd help a lot to keep them all together.

maxime1310 replied:
I'm not sure how to do this because this is specific to the adversarial inference class, I'll see if I can do something about it.

jeff-regier commented on scvi/models/vae_fish.py:
```diff
@@ -0,0 +1,221 @@
+            batch_index = torch.zeros_like(library)
+        if mode == "smFISH":
+            qz_m, qz_v, z = self.z_encoder_fish(x_[:, self.indexes_to_keep])
+            library = torch.log(torch.sum(x[:, self.indexes_to_keep], dim=1)).view(-1, 1)
```
Less branching sounds good

maxime1310 replied:
That seems feasible and should spare some lines of code, I'll try to do that

maxime1310 commented:
@jeff-regier just realized I forgot to bring coverage to its original level. I'm on it.

jeff-regier commented:
@maxime1310 To get rid of these merge conflicts, you could do
```
git fetch origin master
git rebase origin/master
```

Also, how about 
```
mv scRNA_and_smFISH_new_version.ipynb docs/examples/scRNA_and_smFISH.ipynb
```
?
You've got two smFISH notebooks in the pull request currently.

maxime1310 commented:
I had trouble using rebase (I don't know why it took me back to an old version of my branch, several commits ago, and started the conflict resolving from this point), so I just incorporated the two commits on master that have been done since I branched (namely the addition of compute_marginal_log_likelihood method and you're changes for the dense/sparse issue). I also deleted the second notebook, that was a mistake from my part.

jeff-regier commented:
That's how rebase works: it applies your old commits one at at time and each time you fix the conflicts before it proceeds to the next commit.

Maybe `git merge origin/master` would work better in this case because this pull request is so big.

maxime1310 commented:
Yes I think rebase was complicated because there was so many commits, `git merge origin/master` was much faster. All the conflicts must be resolved by now.

jeff-regier commented:
OK. Please make sure the new notebook runs all the way through without errors. You might try running it from the command line like I do below:

https://github.com/YosefLab/scVI/blob/296fdb654c765648747a423ccb614a1531d38dc5/Makefile#L93-L94

jeff-regier commented:
If you put `cd ../..` in a cell by itself at the top of the notebook, it'll get the path right to import scvi.

jeff-regier commented:
Please go ahead and click "Squash and merge" once the command below runs wo error:
```
jupyter nbconvert --to notebook --execute docs/examples/scRNA_and_smFISH.ipynb --ExecutePreprocessor.timeout=1800 
```

(I still get errors even with the latest change.)

maxime1310 commented:
Yes I get the same error for all the notebooks , it must be the path that is not well specified

jeff-regier commented:
The other notebooks should work fine now, after #146

jeff-regier commented:
Maybe merge once more?

maxime1310 commented:
Oh I hadn't seen that change! I was desperatly trying to make the notebook run from the command line  with `os.chdir('../.') ` instead of `cd ../..`. Seems to be working now, I'll just wait for the whole notebook to run

maxime1310 commented:
Testing the notebooks is pretty cool I think!

maxime1310 commented:
The notebook passes the test now! 😄

maxime1310 commented:
Thanks for putting up with this Jeff! Glad that it's finally done!
